### PR TITLE
Feature Random Participants

### LINKED ISSUE #224: Randomized selection of participants is not implemented for IDO
The polkadex IDO pallet allows anyone to create an IDO round. Once that is created, the caller of it can whitelist participants to the IDO round. Also, anyone who is registered as a participant should be able to show_interest in participating in the round. These interested participants should be selected using BABE Randomness. This is not yet implemented in the IDO pallet.

### PR DESCRIPTION
**Pull request by mambisi**
resolves #224 #245 #246

Random Participants,
Motivation for the Algorithm
Algorithm R : https://en.wikipedia.org/wiki/Reservoir_sampling
Participant slots are filled, slot capacity is calculated by max_allocation divided by total amount, then new participants are are placed in random slots

### CODE REVIEW

**Comment by mambisi** [comment]
@Gauthamastro  Please take a look a this PR

File: pallets/Polkadex_Ido/src/lib.rs [hunk 1]
```diff
@@ -182,6 +190,12 @@ impl<T: Config> FundingRound<T> {
     }
 }
 
+#[derive(Decode, Encode, Clone)]
+pub struct InterestedInvestorInfo<T: Config + frame_system::Config> {
+    account_id: T::AccountId,
+    amount: T::Balance,
+}
+
```
Comment (simonsso) [inline pallets/Polkadex_Ido/src/lib.rs:198]: no longer needed remove

File: pallets/Polkadex_Ido/src/lib.rs [hunk 2]
```diff
@@ -390,17 +409,58 @@ decl_module! {
...
             let current_block_no = <frame_system::Pallet<T>>::block_number();
-            let funding_round = <InfoFundingRound<T>>::get(round_id);
             ensure!(current_block_no < funding_round.close_round_block && current_block_no >= funding_round.start_block, <Error<T>>::NotAllowed);
-            InterestedParticipants::<T>::mutate(round_id, |investors| {
-                    investors.push(investor_address.clone());
-                });
-            Self::deposit_event(RawEvent::ShowedInterest(round_id, investor_address));
+
+            InterestedParticipantsAmounts::<T>::mutate(round_id, |interested_participants| {
+                let total_potential_raise : T::Balance = interested_participants.iter()
```
Comment (simonsso) [inline pallets/Polkadex_Ido/src/lib.rs:428]: todo: add ticket for technical dept here. This algo is inefficient

File: pallets/Polkadex_Ido/src/test.rs [hunk 3]
```diff
@@ -271,13 +271,16 @@ fn test_claim_tokens() {
 
 #[test]
 fn test_show_interest_in_round() {
-    let balance: Balance = 100;
+    let balance: Balance = 500;
```
Comment (simonsso) [inline pallets/Polkadex_Ido/src/test.rs:274]: add comments describing the test scenario here.

File: pallets/Polkadex_Ido/src/test.rs [hunk 4]
```diff
@@ -309,13 +314,80 @@ fn test_show_interest_in_round() {
...
+#[test]
+fn test_show_interest_in_round_randomized_participants() {
+    let balance: Balance = 500;
+    let block_num = 3;
+    let min_allocation : Balance = 100;
+    let max_allocation : Balance = 400;
+    ExtBuilder::default().build().execute_with(|| {
         assert_eq!(
-            PolkadexIdo::show_interest_in_round(Origin::signed(investor_address), round_id,),
+            PolkadexIdo::register_round(
```
Comment (simonsso) [inline pallets/Polkadex_Ido/src/test.rs:344]: description here too.

File: pallets/ocex/src/lib.rs [hunk 5]
```diff
@@ -252,18 +252,15 @@ impl<T: Config> Module<T> {
...
-                let index = linked_account
-                    .proxies
-                    .iter()
-                    .position(|x| *x == proxy)
-                    .unwrap();
-                linked_account.proxies.remove(index);
-                Ok(())
-            },
-        )
+        <MainAccounts<T>>::try_mutate(main, |ref mut linked_account: &mut LinkedAccount<T>| {
```
Comment (simonsso) [inline pallets/ocex/src/lib.rs:255]: auto format move to another pr

**Review by simonsso — CHANGES_REQUESTED** [review]
This was done over a voice call with the developer.

File: contracts/uniswap_v2/lib.rs [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+            while i + 1 < path.len() {
+                let (supply_currency_id, target_currency_id) = (path[i], path[i + 1]);
+                let (supply_increment, target_decrement) = (amounts[i], amounts[i + 1]);
+                self._swap(
+                    supply_currency_id,
+                    target_currency_id,
+                    supply_increment,
+                    target_decrement,
+                )?;
+                i += 1;
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:368]: rewrite the loop without the mutable index `i`

!ensure path.len() == amounts.len()

File: contracts/uniswap_v2/lib.rs [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+            min_target_amount: Balance,
+            price_impact_limit: Option<Ratio>,
+        ) -> Result<Balance> {
+            let caller = self.env().caller();
+
+            let amounts = self.get_target_amounts(&path, supply_amount, price_impact_limit)?;
+
+            if amounts[amounts.len() - 1] < min_target_amount {
+                return Err(Error::InsufficientTargetAmount);
+            }
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:386]: If len == 0...

File: contracts/uniswap_v2/lib.rs [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+            &mut self,
+            path: Vec<TokenAddress>,
+            target_amount: Balance,
+            max_supply_amount: Balance,
+            price_impact_limit: Option<Ratio>,
+        ) -> Result<Balance> {
+            let caller = self.env().caller();
+
+            let amounts = self.get_supply_amounts(&path, target_amount, price_impact_limit)?;
+
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:418]: size check

File: contracts/uniswap_v2/lib.rs [hunk 9]
```diff
@@ -0,0 +1,839 @@
...
+        ) -> Option<Balance> {
+            let path_length = path.len();
+            if path_length < 2 || path_length > (TRADING_PATH_LIMIT as usize) {
+                return None;
+            }
+
+            self.get_target_amounts(&path, supply_amount, None)
+                .ok()
+                .map(|amounts| amounts[amounts.len() - 1])
+        }
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:814]: please explain this part.

File: contracts/uniswap_v2/lib.rs [hunk 10]
```diff
@@ -0,0 +1,839 @@
...
+            // price_impact_limit: Option<Ratio>,
+        ) -> Option<Balance> {
+            let path_length = path.len();
+            if path_length < 2 || path_length > (TRADING_PATH_LIMIT as usize) {
+                return None;
+            }
+
+            self.get_supply_amounts(&path, target_amount, None)
+                .ok()
+                .map(|amounts| amounts[0])
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:830]: explain what this is doing

File: pallets/Polkadex_Ido/src/test.rs [hunk 11]
```diff
@@ -271,51 +275,171 @@ fn test_claim_tokens() {
 
 #[test]
 fn test_show_interest_in_round() {
-    let balance: Balance = 100;
-    let investor_address: u64 = 4;
-    let block_num = 3;
+    // Amount to Raise for funding round
```
Comment (simonsso) [inline pallets/Polkadex_Ido/src/test.rs:278]: Overview of test case in comment, no need to duplicate every assignment with an identical comment - give the big picture in a few lines

**Review by simonsso — CHANGES_REQUESTED** [review]
Sanitize input and array lengths

**Comment by simonsso** [comment]
Copy files and make a new branch from develop

**Pull request closed without merging**